File: python/dna_traits.py
"""
A library for parsing and querying 23andme genomes.

Copyright (C) 2014 Christian Stigen Larsen
Distributed under the GPL v3 or later. See COPYING.
"""

import copy
import _dna_traits

def _to_snp(rsid, orientation, value):
    genotype, chromo, position = value
    return SNP(list(genotype), rsid, orientation, chromo, position)

class Nucleotide:
    """A single nucleotide."""
    def __init__(self, value):
        self._value = value

    def __str__(self):
        return self._value

    def __repr__(self):
        return "Nucleotide(%s)" % repr(self._value)

    def __len__(self):
        if self._value in ["-"]:
            return 0
        else:
            return 1

    def __eq__(self, o):
        """Checks if nucleotides are equal."""
        if isinstance(o, Nucleotide):
            return self._value == o._value
        else:
            raise TypeError("Can only compare Nucleotides")

    def complement(self):
        """Returns the complement of this nucleotide."""
        compl = {"A": "T",
                 "C": "G",
                 "G": "C",
                 "T": "A",
                 "D": "D",
                 "I": "I",
                 "-": "-"}
        return Nucleotide(compl[self._value])

    def __invert__(self):
        return self.complement()

class SNP:
    """A single-nucleotide polymorphism."""

    def __init__(self,
                 genotype,
                 rsid,
                 orientation,
                 chromosome,
                 position,
                 phased=False):
        self._genotype = genotype
        self._rsid = rsid
        self._orientation = orientation
        self._chromosome = chromosome
        self._position = position
        self._phased = phased

    def _genostr(self):
        return "".join(map(str, self._genotype))

    @property
    def phased(self):
        """Returns whether this SNP is phased or not."""
        return self._phased

    @property
    def chromosome(self):
        """Returns SNP's chromosome."""
        return self._chromosome

    @property
    def position(self):
        """Returns SNP's position in the chromosome."""
        return self._position

    @property
    def orientation(self):
        """Returns orientation as either -1 or +1."""
        return self._orientation

    def __len__(self):
        return sum(map(len, self._genotype))

    def __getitem__(self, index):
        return self._genotype[index]

    @property
    def rsid(self):
        """Returns this SNP's RSID."""
        return self._rsid

    def count(self, nucleotide):
        """Returns number of given nucleotide in this SNP."""
        return self._genostr().count(nucleotide.upper())

    def complement(self):
        """Returns this SNP's complement."""
        snp = copy.copy(self)
        snp._genotype = map(lambda n: n.complement(), self._genotype)
        return snp

    def positive(self):
        """Returns SNP with positive orientation."""
        if self.orientation < 0:
            return self.complement()
        else:
            return self

    def negative(self):
        """Returns SNP with negative orientation."""
        if self.orientation > 0:
            return self.complement()
        else:
            return self

    def __eq__(self, obj):
        """Rich-comparison operator for SNPs."""
        if isinstance(obj, SNP):
            return str(self.positive()) == str(obj.positive())
        elif isinstance(obj, str):
            if obj[0]=="~":
                # TODO: Should invert obj instead of us; doesn't matter
                genotype = self.negative()._genostr()
                obj = obj[1:]
            else:
                genotype = self.positive()._genostr()

            if not self.phased:
                return genotype==obj or "".join(reversed(genotype))==obj
            else:
                return genotype==obj
        else:
            return NotImplemented

    def __invert__(self):
        return self.complement()

    @property
    def genotype(self):
        """Returns a list of zero, one or two Nucleotides."""
        return self._genotype

    def __str__(self):
        return self._genostr()

    def __repr__(self):
        s = "SNP(genotype=%s, rsid='%s', orientation=%s, " % (
                repr(self._genotype), self._rsid, self._orientation)
        s += "chromosome=%s, position=%s)" % (self._chromosome,
                self._position)
        return s


class GenomeIterator:
    def __init__(self, genome, start=-1):
        self._genome = genome
        self._rsid = genome.first if start==-1 else start

    def __iter__(self):
        return self

    def next(self):
        while self._rsid not in self._genome:
            self._rsid += 1
            if self._rsid >= self._genome.last:
                raise StopIteration()

        rsid = self._rsid
        self._rsid += 1
        return self._genome[rsid]

class Genome:
    """A genome consisting of SNPs."""

    def __init__(self, genome, orientation):
        self._genome = genome
        self._orientation = orientation

    def _rsid(self, rsid):
        if isinstance(rsid, int):
            return rsid
        elif isinstance(rsid, str) and rsid.lower().startswith("rs"):
            try:
                return int(rsid[2:])
            except ValueError:
                raise ValueError("Invalid RSID: %s" % rsid)
        else:
            raise ValueError("Invalid RSID: %s" % rsid)

    def __iter__(self):
        return GenomeIterator(self)

    def intersect_rsid(self, genome):
        """Find RSIDs that exist in both genomes.

        Returns:
            A list of RSID integers.
        """
        assert(isinstance(genome, Genome))
        return sorted(self._genome.intersect_rsid(genome._genome))

    def intersect_snp(self, genome):
        """Find RSID of SNPs that are equal in both genomes.

        Returns:
            A list of RSID integers.
        """
        assert(isinstance(genome, Genome))
        return sorted(self._genome.intersect_snp(genome._genome))

    @property
    def first(self):
        """Returns lowest RSID in genome."""
        return self._genome.first()

    @property
    def last(self):
        """Returns highest RSID in genome."""
        return self._genome.last()

    @property
    def y_chromosome(self):
        """True if genome contains a Y-chromosome."""
        return self._genome.y_chromosome()

    @property
    def male(self):
        """True if genome contains a Y-chromosome."""
        return self.y_chromosome

    @property
    def female(self):
        """False if genome contains a Y-chromosome."""
        return not self.male

    @property
    def orientation(self):
        """Returns this Genome's orientation as an integer of either -1 or
        +1."""
        return self._orientation

    def __getitem__(self, key):
        """Returns SNP with given RSID.  If RSID is not present, return an
        empty SNP."""
        if isinstance(key, slice):
            return [self[i] for i in xrange(*key.indices(len(self)))]
        elif isinstance(key, str) or isinstance(key, int):
            rsid = self._rsid(key)
            try:
                genotype, chromo, position = self._genome[rsid]
                geno = map(Nucleotide, genotype)
                return _to_snp("rs%d" % rsid, self._orientation, (geno, chromo,
                    position))
            except KeyError:
                return SNP([], "rs%d" % rsid, self._orientation, 0, 0)
        else:
            raise ValueError("Unknown key type %s" % type(key))

    def snp(self, rsid):
        """Returns SNP with given RSID."""
        return self.__getitem__(rsid)

    def __str__(self):
        return "Genome"

    def __contains__(self, rsid):
        try:
            self._genome[self._rsid(rsid)]
            return True
        except KeyError:
            return False

    def __repr__(self):
        return "<Genome: SNPs=%d, y_chromosome=%s, orientation=%s>" % (
                len(self), self.y_chromosome, self.orientation)

    def __getattr__(self, rsid):
        # Query with genome.rs28357092
        if isinstance(rsid, str) and rsid.startswith("rs"):
            return self.__getitem__(rsid)
        raise AttributeError("Unknown attribute %s" % rsid)

    def __len__(self):
        """Returns number of SNPs in this genome."""
        return len(self._genome)

    def __eq__(self, genome):
        return self._genome.eq(genome._genome)

    def __ne__(self, genome):
        return not self.__eq__(genome)

    def match(self, criteria):
        """Match list of (RSID, BasePair) with genome. BasePair should be a
        string with positive orientation.

        Example:
            all(match(genome, [("rs4778241", "CC"), ("rs1291832", "GG")]))
        """
        return (self[rsid]==genotype for rsid, genotype in criteria)


def parse(filename, orientation=+1):
    """Parses a 23andme file and returns a Genome."""
    return Genome(_dna_traits.parse(filename), orientation)

File: dnatraits/src/dnatraits.cpp
/*
 * Copyright (C) 2014 Christian Stigen Larsen
 * Distributed under the GPL v3 or later. See COPYING.
 */

#include <sstream>
#include <google/dense_hash_map>

#include "dnatraits.hpp"

struct DLL_LOCAL RSIDHash {
  inline std::size_t operator() (const RSID& rsid) const
  {
    return static_cast<std::size_t>(rsid);
  }
};

struct DLL_LOCAL RSIDEq {
  inline bool operator()(const RSID& a, const RSID& b) const
  {
    return a == b;
  }
};

typedef google::dense_hash_map<RSID, SNP, RSIDHash, RSIDEq> SNPMap;

const Genotype AA (A, A);
const Genotype AC (A, C);
const Genotype AG (A, G);
const Genotype AT (A, T);
const Genotype CA (C, A);
const Genotype CC (C, C);
const Genotype CG (C, G);
const Genotype CT (C, T);
const Genotype GA (G, A);
const Genotype GC (G, C);
const Genotype GG (G, G);
const Genotype GT (G, T);
const Genotype NN (NONE, NONE);
const Genotype TA (T, A);
const Genotype TC (T, C);
const Genotype TG (T, G);
const Genotype TT (T, T);
const SNP NONE_SNP(NO_CHR, 0, NN);

std::ostream& operator<<(std::ostream& o, const Nucleotide& n)
{
  switch ( n ) {
    case A: return o << 'A';
    case C: return o << 'C';
    case D: return o << 'D';
    case G: return o << 'G';
    case I: return o << 'I';
    case NONE: return o << '-';
    case T: return o << 'T';
  }
}

std::ostream& operator<<(std::ostream& o, const Genotype& bp)
{
  return o << bp.first << bp.second;
}

std::ostream& operator<<(std::ostream& o, const Chromosome& chr) {
  if ( chr >= NO_CHR && chr < CHR_MT )
    return o << static_cast<int>(chr);

  switch ( chr ) {
    default: return o;
    case CHR_MT: return o << "MT";
    case CHR_X: return o << "X";
    case CHR_Y: return o << "Y";
  }
}

std::ostream& operator<<(std::ostream& o, const SNP& snp)
{
  return o << snp.genotype << " " << snp.chromosome
    << " " << snp.position;
}

/**
 * Returns the complement nucleotide.
 */
Nucleotide complement(const Nucleotide& n)
{
  switch ( n ) {
    case A: return T;
    case C: return G;
    case G: return C;
    case T: return A;
    case D: return D;
    case I: return I;
    case NONE: return NONE;
  }
}

Genotype::Genotype(const Nucleotide& a, const Nucleotide& b)
  : first(a), second(b)
{
}

Genotype operator~(const Genotype& g)
{
  return Genotype(complement(g.first),
                  complement(g.second));
}

bool Genotype::operator==(const Genotype& g) const
{
  return first == g.first && second == g.second;
}

bool Genotype::operator<(const Genotype& g) const
{
  if ( first < g.first )
    return true;

  if ( first > g.first )
    return false;

  return second < g.second;
}

SNP::SNP(const Chromosome& chr,
    const Position& pos,
    const Genotype& gt) :
  chromosome(chr),
  position(pos),
  genotype(gt)
{
}

SNP::SNP(const SNP& snp) :
  chromosome(snp.chromosome),
  position(snp.position),
  genotype(snp.genotype)
{
}

SNP& SNP::operator=(const SNP& snp) {
  if ( this != &snp ) {
    genotype = snp.genotype;
    chromosome = snp.chromosome;
    position = snp.position;
  }
  return *this;
}

bool SNP::operator==(const SNP& snp) const
{
  return genotype == snp.genotype &&
         chromosome == snp.chromosome &&
         position == snp.position;
}

bool SNP::operator<(const SNP& snp) const
{
  if ( position > snp.position )
    return false;
  if ( position < snp.position )
    return true;

  // equal position
  if ( chromosome > snp.chromosome )
    return false;
  if ( chromosome < snp.chromosome )
    return true;

  // equal chromosome
  return genotype < snp.genotype;
}

bool SNP::operator>(const SNP& snp) const
{
  return !(*this <= snp);
}

bool SNP::operator<=(const SNP& snp) const
{
  return *this == snp || *this < snp;
}

bool SNP::operator>=(const SNP& snp) const
{
  return *this == snp || *this > snp;
}

bool SNP::operator!=(const SNP& snp) const
{
  return !(*this == snp);
}

bool SNP::operator==(const Genotype& g) const
{
  return genotype == g;
}

struct DLL_LOCAL Genome::GenomeImpl {
  DLL_LOCAL SNPMap snps;

  GenomeImpl(const size_t size) :
    snps(size)
  {
    snps.set_empty_key(0);
  }

  GenomeImpl(const GenomeImpl& g) :
    snps(g.snps)
  {
    snps.set_empty_key(0);
  }

  GenomeImpl& operator=(const GenomeImpl& g)
  {
    if ( this != &g )
      snps = g.snps;

    return *this;
  }

  bool contains(const RSID& rsid) const {
    return snps.find(rsid) != snps.end();
  }

  const SNP& operator[](const RSID& rsid) const {
    return !contains(rsid)? NONE_SNP : const_cast<SNPMap&>(snps)[rsid];
  }
};


Genome::Genome(const size_t size):
  y_chromosome(false),
  first(0xffffffff),
  last(0),
  pimpl(new GenomeImpl(size))
{
}

Genome::Genome(const Genome& g) :
  y_chromosome(g.y_chromosome),
  first(g.first),
  last(g.last),
  pimpl(new GenomeImpl(*g.pimpl))
{
}

Genome& Genome::operator=(const Genome& g)
{
  if ( this != &g ) {
    *pimpl = *g.pimpl;
    y_chromosome = g.y_chromosome;
    first = g.first;
    last = g.last;
  }
  return *this;
}

Genome::~Genome()
{
  delete pimpl;
}

const SNP& Genome::operator[](const RSID& rsid) const
{
  return (*pimpl)[rsid];
}

bool Genome::has(const RSID& rsid) const
{
  return pimpl->contains(rsid);
}

size_t Genome::size() const
{
  return pimpl->snps.size();
}

double Genome::load_factor() const
{
  return pimpl->snps.load_factor();
}

void Genome::insert(const RSID& rsid, const SNP& snp)
{
  pimpl->snps.insert({rsid, snp});
}

std::vector<RSID> Genome::intersect_rsid(const Genome& genome) const
{
  std::vector<RSID> r;

  for ( const auto i : pimpl->snps )
    if ( genome.has(i.first) )
      r.push_back(i.first);

  return r;
}

std::vector<RSID> Genome::intersect_snp(const Genome& genome) const
{
  std::vector<RSID> r;

  for ( const auto i : pimpl->snps )
    if ( genome.has(i.first) )
      if ( genome[i.first] == operator[](i.first) )
        r.push_back(i.first);

  return r;
}

std::vector<RSID> Genome::rsids() const
{
  std::vector<RSID> r;

  for ( const auto i : pimpl->snps )
    r.push_back(i.first);

  return r;
}

bool Genome::operator==(const Genome& o) const
{
  if ( first != o.first ) return false;
  if ( last != o.last ) return false;
  if ( y_chromosome != o.y_chromosome ) return false;
  if ( o.size() != size() ) return false;

  for ( const auto i : pimpl->snps ) {
    if ( !o.has(i.first) )
      return false;

    if ( i.second != o[i.first] )
      return false;
  }

  return true;
}

bool Genome::operator!=(const Genome& o) const
{
  return !(*this == o);
}
